Fix drag and drop support on linux (and macOS?) with SDL and .NET core

morguldir opened the pull request:
### Purpose of this PR

Make Drag and drop support work when running on non-unicode platforms with SDL because .NET Core ~was too lazy to do it~ handles text encoding differently.

Resolves https://github.com/ppy/osu/issues/3444

morguldir committed:
Fix drag and drop support on linux (and macOS?) with SDL and .NET core

peppy commented:
Thanks for submitting this. Any idea how it regressed, since it seems to have been working up to a certain point.

morguldir commented:
It was working when SDL wasn’t loading and we were falling back to X11, which #27 fixed

VPeruS commented on src/OpenTK/Platform/SDL2/Sdl2NativeWindow.cs:
```diff
@@ -303,9 +303,16 @@ private static void ProcessMouseWheelEvent(Sdl2NativeWindow window, MouseWheelEv
             window.OnMouseWheel(ev.X, ev.Y, false);
         }
 
+        private static readonly int SystemDefaultCharSize = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? 2 : 1;
```
```suggestion
        private static readonly bool IsFilenameUniEncoded = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
```

VPeruS replied:
Move it to the beginning of class declaration

VPeruS commented on src/OpenTK/Platform/SDL2/Sdl2NativeWindow.cs:
```diff
@@ -303,9 +303,16 @@ private static void ProcessMouseWheelEvent(Sdl2NativeWindow window, MouseWheelEv
+
+        private static string PtrToStringAuto(IntPtr ptr)
+        {
+            return SystemDefaultCharSize == 2 ? Marshal.PtrToStringUni(ptr) : Marshal.PtrToStringAnsi(ptr);
```
```suggestion
            => IsFilenameUniEncoded ? Marshal.PtrToStringUni(ptr) : Marshal.PtrToStringAnsi(ptr);
```

VPeruS commented:
Fixed review comments

morguldir committed:
src/OpenTK/Platform/SDL2/Sdl2NativeWindow.cs

Remove private SystemDefaultCharSize and only use IsOSPlatform
Use expression body definition for private PtrToStringAuto function

VPeruS commented:
Try to test with utf-8 encoded filenames like "my-𝌆-file"

morguldir commented:
![1540236477](https://user-images.githubusercontent.com/21371374/47315026-ae496a80-d643-11e8-9772-44d75faef301.png)

morguldir committed:
Merge branch 'netstandard' of https://github.com/ppy/opentk into fix_drag_drop

VPeruS commented on src/osuTK/Platform/SDL2/Sdl2NativeWindow.cs:
```diff
@@ -303,9 +303,12 @@ private static void ProcessMouseWheelEvent(Sdl2NativeWindow window, MouseWheelEv
         }
 
+        private static string PtrToStringAuto(IntPtr ptr)
+        => RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? Marshal.PtrToStringUni(ptr) : Marshal.PtrToStringAnsi(ptr);
```
```suggestion
            => RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? Marshal.PtrToStringUni(ptr) : Marshal.PtrToStringAnsi(ptr);
```

VPeruS committed:
Put 4 spaces before => operator

Co-Authored-By: morguldir <morguldir@protonmail.com>

morguldir committed:
Merge branch 'netstandard' into fix_drag_drop

smoogipoo committed:
Merge branch 'netstandard' into fix_drag_drop

smoogipoo commented:
Tested on Windows and macOS, seems to work.
